ykrt/src/compile/jitc_yk/codegen/x64/mod.rs

```diff
@@ -419,13 +419,79 @@ impl<'a> Assemble<'a> {
 
         match inst.binop() {
             BinOp::Add => {
-                let size = lhs.byte_size(self.m);
+                let byte_size = lhs.byte_size(self.m);
+                match (&lhs, &rhs) {
+                    (Operand::Const(cidx), Operand::Var(_))
+                    | (Operand::Var(_), Operand::Const(cidx)) => {
+                        let Const::Int(ctyidx, v) = self.m.const_(*cidx) else {
+                            unreachable!()
+                        };
+                        let Ty::Integer(bit_size) = self.m.type_(*ctyidx) else {
+                            panic!();
+                        };
+                        // Try to emit `add` instructions with immediate operands.
+                        //
+                        // We assume nothing about the unused upper bits (if any) of `lhs_reg` and
+                        // the `u64` backing the constant, so we can't simply try and use `add r64,
+                        // imm32` to cover lots of cases. This would lead to undesired overflows
+                        // caused by (potentially set) upper bits.
+                        let mut get_lhs_reg = |lhs| {
+                            self.ra.assign_gp_regs(
+                                &mut self.asm,
+                                iidx,
+                                [RegConstraint::InputOutput(lhs)],
+                            )
+                        };
+                        match bit_size {
+                            8 => {
+                                if let Ok(_v8) = i8::try_from(*v as i64) {
```

Comment:
Are we sure that this check always does what we want? I haven't worked through the edge cases, so it *might* be OK. I kinda assumed we had to do truncation/sign extension ourselves, but maybe I'm wrong!